File: source/_posts/Algorithm/数据结构/二叉树/二叉搜索树/95.不同的二叉搜索树-ii.py
#
# @lc app=leetcode.cn id=95 lang=python3
#
# [95] 不同的二叉搜索树 II
#  BST

# @lc code=start
# Definition for a binary tree node.
from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# 我们将问题进一步划分为子问题，假如左侧和右侧的树分别求好了，
# 我们是不是只要运用组合的原理，将左右两者进行合并就好了，我们需要两层循环来完成这个过程。


class Solution:
    def generateTrees(self, n: int) -> List[TreeNode]:
        if not n:
            return []

        def generateTree(start, end):
            if start > end:
                return [None]
            res = []
            for i in range(start, end + 1):
                ls = generateTree(start, i - 1)
                rs = generateTree(i + 1, end)
                for l in ls:
                    for r in rs:
                        node = TreeNode(i)
                        node.left = l
                        node.right = r
                        res.append(node)
            return res

        return generateTree(1, n)


# @lc code=end
test = Solution()
res = test.generateTrees(3)
print("e")

File: source/_posts/Algorithm/数据结构/二叉树/二叉搜索树/95.不同的二叉搜索树-ii.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

import javax.xml.stream.events.EntityDeclaration;

import apple.laf.JRSUIUtils.Tree;

/*
 * @lc app=leetcode.cn id=95 lang=java
 *
 * [95] 不同的二叉搜索树 II
 */

// @lc code=start
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode() {} TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) { this.val = val; this.left
 * = left; this.right = right; } }
 */
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public List<TreeNode> generateTrees(int n) {
        return gen(1, n);
    }

    private List<TreeNode> gen(int start, int end) {
        LinkedList<TreeNode> res = new LinkedList<TreeNode>();
        if (start > end) {
            res.add(null);
            return res;
        }
        for (int i = start; i <= end; i++) {
            List<TreeNode> leftTrees = gen(start, i - 1);
            List<TreeNode> rightTrees = gen(i + 1, end);
            for (TreeNode left : leftTrees) {
                for (TreeNode right : rightTrees) {
                    TreeNode currTree = new TreeNode(i);
                    currTree.left = left;
                    currTree.right = right;
                    res.add(currTree);
                }
            }
        }
        return res;
    }

    // public static void main(String[] args) {
    // Solution95 solution95 = new Solution95();
    // solution95.generateTrees(3);
    // }
}
// @lc code=end
